File: leetcode/207. Course Schedule.cpp
// Copyright 2016 Qi Wang
// Date: 2016-10-22
class Solution {
 public:
  bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
    vector<list<int>> map(numCourses);
    vector<int> inDegree(numCourses, 0);
    for (auto& p : prerequisites) {
      map[p.first].push_back(p.second);
      ++inDegree[p.second];
    }
    queue<int> zeroInDegreeQ;
    for (int i = 0; i < numCourses; ++i)
      if (inDegree[i] == 0) zeroInDegreeQ.push(i);
    while (!zeroInDegreeQ.empty()) {
      int u = zeroInDegreeQ.front();
      zeroInDegreeQ.pop();
      for (int v : map[u])
        if (--inDegree[v] == 0) zeroInDegreeQ.push(v);
    }
    return all_of(inDegree.begin(), inDegree.end(), [](int degree) {
      return degree == 0;
    });
  }
};

// Date: 2015-08
class Solution
{
public:
	bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites)
	{
		vector<vector<bool>> map(numCourses, vector<bool>(numCourses, false));
		vector<int> in_degree(numCourses, 0);
		for_each(prerequisites.begin(), prerequisites.end(),
			[&map, &in_degree](pair<int,int> p)
			{
				if (map[p.first][p.second] == false)
				{
					map[p.first][p.second] = true;
					++ in_degree[p.second];
				}
			});
		queue<int> q;
		for (int id = 0; id < numCourses; ++ id)
		{
			if (in_degree[id] == 0)
			{
				q.push(id);
			}
		}
		while (!q.empty())
		{
			int course = q.front();
			q.pop();
			for (int id = 0; id < numCourses; ++ id)
			{
				if (map[course][id] == true)
				{
					-- in_degree[id];
					if (in_degree[id] == 0)
					{
						q.push(id);
					}
				}
			}
		}
		return all_of(in_degree.begin(), in_degree.end(), [](int in) -> bool
		{
			return in == 0;
		});
	}
};

File: leetcode/207. Course Schedule.java
// Date: 2020-09-07
// Refer to:
//   - http://www.personal.kent.edu/~rmuhamma/Algorithms/MyAlgorithms/GraphAlgor/strongComponent.htm
//   - https://leetcode.com/problems/course-schedule/discuss/249688/Different-O(V%2BE)-solution-using-Kosaraju's-algorithm
class Solution {
  public boolean canFinish(int numCourses, int[][] prerequisites) {
    List<Integer>[] g = new ArrayList[numCourses];
    List<Integer>[] rg = new ArrayList[numCourses];
    for (int i = 0; i < numCourses; ++i) {
      g[i] = new ArrayList<Integer>();
      rg[i] = new ArrayList<Integer>();
    }
    for (int[] pre : prerequisites) {
      g[pre[0]].add(pre[1]);
      rg[pre[1]].add(pre[0]);
    }
    boolean[] visited = new boolean[numCourses];
    Stack<Integer> stk = new Stack<Integer>();
    // 1st dfs on g
    for (int i = 0; i < numCourses; ++i) {
      if (!visited[i]) {
        dfs(i, g, visited, stk);
      }
    }
    // 2nd dfs on reversed graph, rg
    Arrays.fill(visited, false);
    while (!stk.isEmpty()) {
      int u = stk.pop();
      if (!visited[u] && dfs(u, rg, visited) > 1) {
        return false;
      }
    }
    return true;
  }
  
  private void dfs(int u, List<Integer>[] g, boolean[] visited, Stack<Integer> stk) {
    visited[u] = true;
    g[u].stream()
        .filter(v -> !visited[v])
        .forEach(v -> dfs(v, g, visited, stk));
    stk.push(u);
  }
  
  private int dfs(int u, List<Integer>[] g, boolean[] visited) {
    visited[u] = true;
    return g[u].stream()
        .filter(v -> !visited[v])
        .reduce(1, (acc, v) -> acc + dfs(v, g, visited));
  }
}